packages/runtime/container-runtime/src/orderedClientElection.ts

```diff
@@ -0,0 +1,263 @@
+            // Update linked list end pointer to youngest client.
+            this.youngestClient = newClient;
+        } else {
+            // Update next node to point back to this new node.
+            newClient.youngerClient.olderClient = newClient;
+        }
+
+        this.eligibleCount++;
+        return newClient;
+    }
+
+    private deleteEligibleClient(removeClient: Readonly<IEligibleClient>) {
+        // Update prev node to point to next node.
+        removeClient.olderClient.youngerClient = removeClient.youngerClient;
+
+        if (removeClient.youngerClient === undefined) {
+            // Update linked list end pointer to youngest client.
+            this.youngestClient = removeClient.olderClient;
+        } else {
+            // Update next node to point back to previous node.
+            removeClient.youngerClient.olderClient = removeClient.olderClient;
+        }
+
+        this.eligibleCount--;
+    }
+
+    private readonly addClient = (clientId: string, client: ISequencedClient) => {
+        const isSummarizer = client.client.details?.type === summarizerClientType;
+        const eligible = !isSummarizer && (client.client.details?.capabilities.interactive ?? true);
+        const newClient: TrackedClient = eligible ? this.insertEligibleClient(clientId, client.sequenceNumber) : {
+            clientId,
+            sequenceNumber: client.sequenceNumber,
+            eligible,
+            isSummarizer,
+        };
+        this.clientMap.set(clientId, newClient);
+
+        // Emit change events if necessary
+        if (newClient.eligible) {
+            if (this.electedClient === undefined && newClient.youngerClient === undefined) {
+                this.electedClient = newClient;
+                this.emit("electedChange", this.getElectedClient());
+            }
+        } else {
+            if (newClient.isSummarizer) {
+                this.summarizerCount++;
+                this.emit("summarizerChange", this.summarizerCount);
+            }
+        }
+    };
+
+    private readonly removeClient = (clientId: string) => {
+        const removeClient = this.clientMap.get(clientId);
+        if (removeClient !== undefined) {
+            this.clientMap.delete(clientId);
+            if (!removeClient.eligible) {
+                if (removeClient.isSummarizer) {
+                    this.summarizerCount--;
+                    this.emit("summarizerChange", this.summarizerCount);
+                }
+                return;
+            }
+
+            this.deleteEligibleClient(removeClient);
+            if (removeClient === this.electedClient) {
+                this.electedClient = this.electedClient.youngerClient;
+                this.emit("electedChange", this.getElectedClient());
+            }
+        }
+    };
+
+    /** Returns the currently elected client. */
+    public getElectedClient(): ITrackedClient | undefined {
+        return this.electedClient;
```

Comment:
should this ever return root, seems like it can now. i think in that case it should return undefined

Reply:
It shouldn't and cannot right now. It only ever gets set to `youngerClient` or `newClient`. And no client has the root as its `youngerClient`.

1. Add client will set it to the added client in some cases (new)
2. Remove client will set it to the younger client (next)
3. Increment will set it to the younger client (next)
4. Reset will set it to the root's younger client (next)